ttl: Add some tools to build SQLs for TTL

lcwangchao opened the pull request:
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: close #39339

### What is changed and how it works?

Add some tools to generate SQL of the TTL:

- Add `ttl.SQLBuilder` to provide some common methods to build a SELECT/DELETE SQL
- Add `ttl.ScanQueryGenerator` to generate scan task SQLs.

### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

lcwangchao commented:
/run-unit-test

YangKeao reviewed (approved):
LGTM

xhebox commented on ttl/sql.go:
```diff
@@ -0,0 +1,468 @@
+// SQLBuilder is used to build SQLs for TTL
+type SQLBuilder struct {
+	tbl   *PhysicalTable
+	sb    strings.Builder
```
Maybe you should consider https://github.com/pingcap/tidb/blob/d100e933c8982ed13135abbca1224b472b167b53/util/sqlexec/utils.go#L247
if there is a risk of injection.

lcwangchao replied:
I'm using `sqlexec.EscapeString` to escape the value. Seems they are equivalent in implement

xhebox replied:
Sure , but table name, or every identifier may have a problem. BTW, `FormatSQL` is the builder version of `EscapeSQL`. Check again if you really need escaping everywhere or just datum.

lcwangchao commented:
> Sure , but table name, or every identifier may have a problem. BTW, FormatSQL is the builder version of EscapeSQL. Check again if you really need escaping everywhere or just datum.

@xhebox Yes, you are right. We still to escape table, partition and column names. I've updated the PR and added some tests.

hawkingrei commented on ttl/BUILD.bazel:
```diff
@@ -0,0 +1,41 @@
+    ],
+)
+
+go_test(
+    name = "ttl_test",
+    srcs = [
+        "main_test.go",
+        "sql_test.go",
+    ],
+    deps = [
+        ":ttl",
+        "//kv",
+        "//parser/model",
+        "//parser/mysql",
+        "//testkit",
+        "//testkit/testsetup",
+        "//types",
+        "//util/sqlexec",
+        "@com_github_stretchr_testify//require",
+        "@org_uber_go_goleak//:goleak",
+    ],
+)
```
```suggestion
go_test(
    name = "ttl_test",
    srcs = [
        "main_test.go",
        "sql_test.go",
    ],
    deps = [
        ":ttl",
        "//kv",
        "//parser/model",
        "//parser/mysql",
        "//testkit",
        "//testkit/testsetup",
        "//types",
        "//util/sqlexec",
        "@com_github_stretchr_testify//require",
        "@org_uber_go_goleak//:goleak",
    ],
    flaky = True,
)
```

hawkingrei replied:
add ```flaky = True,```

hawkingrei commented:
/merge

hawkingrei commented:
/merge
